### PR TITLE
CDAP-8708 Proper owner verification

### PR DESCRIPTION
**Pull request by rohitsinha54**
Do proper verification of the owner. See: https://github.com/caskdata/cdap/compare/bugfix_release/CDAP-8708-owner-fix?expand=1#diff-e3a6b473eb2408f2f77a1c5b6dc34e2fR252

Build: http://builds.cask.co/browse/CDAP-RUT699-5
ITM: http://builds.cask.co/browse/CDAP-ITM11-9

### CODE REVIEW

File: cdap-common/src/main/java/co/cask/cdap/common/kerberos/SecurityUtil.java [hunk 1]
```diff
@@ -245,4 +247,33 @@ public static ImpersonationInfo createImpersonationInfo(OwnerAdmin ownerAdmin, C
     }
     return impersonationInfo;
   }
+
+  /**
+   * <p>Verifies the owner principal of an entities is same as the owner specified during entity creation. If an owner
```
Comment (Nishith) [inline cdap-common/src/main/java/co/cask/cdap/common/kerberos/SecurityUtil.java]: s/entities/entity/

File: cdap-common/src/main/java/co/cask/cdap/common/kerberos/SecurityUtil.java [hunk 2]
```diff
@@ -245,4 +247,33 @@ public static ImpersonationInfo createImpersonationInfo(OwnerAdmin ownerAdmin, C
...
+   * @throws ConflictException if the specified owner information is not valid with the existing
+   * impersonation principal
+   */
+  public static void verifyOwnerPrincipal(NamespacedEntityId existingEntity,
+                                    @Nullable String specifiedOwnerPrincipal,
+                                    OwnerAdmin ownerAdmin) throws IOException, ConflictException {
+    // if an owner principal was not specified then ensure that a direct owner doesn't exists. Although, if an owner
+    // principal was specified then it must be equals to the effective impersonating principal of this entity
+    if (!((specifiedOwnerPrincipal == null && ownerAdmin.getOwnerPrincipal(existingEntity) == null) ||
+      Objects.equals(specifiedOwnerPrincipal, ownerAdmin.getImpersonationPrincipal(existingEntity)))) {
```
Comment (Nishith) [inline cdap-common/src/main/java/co/cask/cdap/common/kerberos/SecurityUtil.java]: Wouldn't `ownerAdmin.getImpersonationPrincipal(existingEntity)` return null if the namespace has no owner? If the user specifies cdap as the specifiedOwnerPrincipal for an entry that was not created with an owner in a namespace without an owner it should still succeed.

File: cdap-common/src/main/java/co/cask/cdap/common/kerberos/SecurityUtil.java [hunk 3]
```diff
@@ -245,4 +247,33 @@ public static ImpersonationInfo createImpersonationInfo(OwnerAdmin ownerAdmin, C
...
+   * @param specifiedOwnerPrincipal the specified principal
+   * @param ownerAdmin {@link OwnerAdmin}
+   * @throws IOException  if failed to query the given ownerAdmin
+   * @throws ConflictException if the specified owner information is not valid with the existing
+   * impersonation principal
+   */
+  public static void verifyOwnerPrincipal(NamespacedEntityId existingEntity,
+                                    @Nullable String specifiedOwnerPrincipal,
+                                    OwnerAdmin ownerAdmin) throws IOException, ConflictException {
+    // if an owner principal was not specified then ensure that a direct owner doesn't exists. Although, if an owner
```
Comment (Nishith) [inline cdap-common/src/main/java/co/cask/cdap/common/kerberos/SecurityUtil.java]: s/exists/exist/, s/equals/equal/

File: cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/services/ApplicationLifecycleServiceTest.java [hunk 4]
```diff
@@ -152,6 +154,42 @@ public void testOwner() throws Exception {
   }
 
   @Test
+  public void testOwnerInImpersonatedNS() throws Exception {
```
Comment (Nishith) [inline cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/services/ApplicationLifecycleServiceTest.java:157]: Add a test without a namespace owner? App impersonation should work in a namespace without an explicitly set principal too.

Comment (Nishith) [inline cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/services/ApplicationLifecycleServiceTest.java:157]: Perhaps add a test for an app with an owner and in an impersonated namespace?

**Review by anew — COMMENTED** [review]
The changes look good.

But I have one remark. As we discussed in person, if the new and old owner were the same, then these operations would succeed and replace the existing entity. So it appears that this is really a lack of privilege by a different user to overwrite this entity. Would it make more sense to return 403 Unauthorized instead 409 Conflict? I am not sure what we agreed on...

File: cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/services/ApplicationLifecycleServiceTest.java (hunk 4 above)
Comment (rohitsinha54) [reply cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/services/ApplicationLifecycleServiceTest.java:157]: that test is already there

Comment (Nishith) [reply cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/services/ApplicationLifecycleServiceTest.java:157]: 👍

**Comment by rohitsinha54** [comment]
@anew I will switch it Unauthorized.

**Comment by anew** [comment]
LGTM. Do we need to update the docs?

**Comment by Nishith** [comment]
LGTM.

**Pull request merged**